File: src_erl/main_5.erl
-module(main_5).

-export([main/0]).


-spec main() -> {ok, books_shop:order()} | {error, term()}.
main() ->
    handle_create_order(books_shop:test_data()).


-spec handle_create_order(map()) -> {ok, books_shop:order()} | {error, term()}.
handle_create_order(Data) ->
    wannabe_haskell:do([
        {data, fun books_shop:validate_incoming_data/1, Data},

        {cat_0, fun(#{<<"cat">> := Cat0}) -> {ok, Cat0} end, data},
        {cat, fun books_shop:validate_cat/1, cat_0},

        {address_0, fun(#{<<"address">> := Address0}) -> {ok, Address0} end, data},
        {address, fun books_shop:validate_address/1, address_0},

        {books_0, fun(#{<<"books">> := Books0}) -> {ok, Books0} end, data},
        {books,
            fun(Books0) ->
                pipeline:sequence(
                    lists:map(
                        fun(#{<<"title">> := Title, <<"author">> := Author}) ->
                            books_shop:get_book(Title, Author)
                        end,
                        Books0
                    ))
            end,
            books_0},

        {order, fun books_shop:create_order/3, [cat, address, books]}
    ]).

File: src_ex/Main5.ex
defmodule Main5 do

    @spec main :: {:ok, BookShop.Order.t} | {:error, term}
    def main do
        BookShop.test_data |> handle_create_order
    end


    @spec handle_create_order(map) :: {:ok, BookShop.Order.t} | {:error, term}
    def handle_create_order data0 do
        with {:ok, data} <- BookShop.validate_incoming_data(data0),
             %{"cat" => cat0, "address" => address0, "books" => books0} = data,
             {:ok, cat} <- BookShop.validate_cat(cat0),
             {:ok, address} <- BookShop.validate_address(address0),
             books1 = Enum.map(books0,
                 fn %{"title" => title, "author" => author} ->
                     BookShop.get_book title, author
                 end),
             {:ok, books2} <- Pipeline.sequence(books1),
             do: {:ok, BookShop.create_order(cat, address, books2)}
    end

end
